ambry-quota/src/main/java/com/github/ambry/quota/capacityunit/AmbryCUQuotaEnforcer.java

```diff
@@ -113,9 +116,9 @@ public void shutdown() {
    * @return QuotaRecommendation object.
    */
   private QuotaRecommendation buildQuotaRecommendation(float usage, QuotaName quotaName) {
-    boolean shouldThrottle = (usage >= 100);
-    return new QuotaRecommendation(shouldThrottle, usage, quotaName,
-        shouldThrottle ? throttleRetryAfterMs : NO_THROTTLE_RETRY_AFTER_MS);
+    QuotaAction quotaAction = (usage >= 100) ? QuotaAction.DELAY : QuotaAction.ALLOW;
+    return new QuotaRecommendation(quotaAction, usage, quotaName,
+        (quotaAction == QuotaAction.DELAY) ? throttleRetryAfterMs : NO_THROTTLE_RETRY_AFTER_MS);
```

Comment:
nit: I feel like the concept need to be consistent. From this logic, when quotaAction != DELAY, which could be ACTIVE or REJECT, then the getRetryAfterMs() will return -1. And the java doc in QuotaRecommendation.java mentioned "If request is not throttled then returns -1". But when I check this logic:
`  public boolean shouldThrottle() {
    return this.quotaAction == QuotaAction.REJECT || this.quotaAction == QuotaAction.DELAY;
  }`
It seems like REJECT and DELAY will both to be considered as should throttle.

Reply:
You are right that both REJECT and DELAY will be considered as "throttled", because both of them means that the request is out of quota and needs to be controlled. "REJECT" and "DELAY" are the recommended action for controlling.
Not every QuotaEnforcer needs to send both. In the current implementation, requests that are out of quota for bandwidth will never be rejected, but will be delayed. Hence CU implementations will set action as "DELAY" for out-of-quota requests. The requests that are out of quota for storage quota will be rejected. Hence storage quota sets "REJECT" as action and not "DELAY". In future there could be some fancy enforcer implementation that can set DELAY in certain conditions, while REJECT in certain other conditions (maybe something like DELAY for bandwidth out-of-quota, but if there are a lot of requests in queue, then REJECT).

So overall both REJECT and DELAY means that request is out-of-quota and needs to be controlled. 
We will only set NO_THROTTLE_RETRY_AFTER_MS if request is not throttled, i.e, if request is allowed.